src/Generator/Shared/CircularReferenceChecker.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AutoMapper\Generator\Shared;
+
+use AutoMapper\MapperMetadata\MapperGeneratorMetadataInterface;
+use AutoMapper\MapperMetadata\MapperGeneratorMetadataRegistryInterface;
+use AutoMapper\Transformer\DependentTransformerInterface;
+
+final readonly class CircularReferenceChecker
```

Comment:
this have been extracted from `MapperMetadata` 

I'd like to de-complexify some objects, this is a part of it